EVG-17602, EVG-17621: Parallelize and independently write to S3

julianedwards opened the pull request:
JIRA: https://jira.mongodb.org/browse/EVG-17602, https://jira.mongodb.org/browse/EVG-17621

I made the writes to S3 as independent as possible but we still rely on the `s3` field from the build model when uploading logs and test metadata. This isn't a big deal, though, because when we are ready to remove the DB code we can just ignore it since all writes will go to S3.

I tested that this works in staging.

ybrill commented:
Did it end up to not be worthwhile to do the S3 and DB writes in parallel?
If they aren't parallel does it matter which is first?

julianedwards commented:
> Did it end up to not be worthwhile to do the S3 and DB writes in parallel? If they aren't parallel does it matter which is first?

It didn't seem worthwhile since we rely on the DB look ups to fetch metadata for now. I think it is still worth doing the S3 writes before the DB writes though (so those are going to be the slowest operations).

ybrill commented:
But I mean, couldn't we do whatever the plan will be for the post-DB era to avoid relying on the db lookups?

julianedwards commented:
I created https://jira.mongodb.org/browse/EVG-17621 for that work since it isn't as straightforward.

ybrill commented:
Ah, I understand now.
So about
> doing the S3 writes before the DB writes though (since those are going to be the slowest operations).

If the end result of the write is going to be a 500 error because of the db write taking forever why is it important to make sure the S3 write succeeds?

julianedwards commented:
> If the end result of the write is going to be a 500 error because of the db write taking forever why is it important to make sure the S3 write succeeds?

The DB writes don't actually use the context (old driver). So when there is high traffic and the DB slows down, the writes take much longer and by the time we get to the S3 write (which uses the context) it hits the minute timeout. We can assume that the S3 writes will be fast. If we want to be extra paranoid we can parallelize the DB and S3 writes, but it will still have to come after the DB lookups until after [EVG-17621](https://jira.mongodb.org/browse/EVG-17621).

johndaniels commented on storage/persistence.go:
```diff
@@ -18,6 +20,8 @@ func (b *Bucket) UploadBuildMetadata(ctx context.Context, build model.Build) err
 	return errors.Wrapf(b.Put(ctx, metadata.key(), bytes.NewReader(json)), "putting metadata for build '%s'", build.Id)
 }
 
+// UploadBuildMetadata uploads metadata for a new test to the offline blob
```
what do we mean by "offline" blob bucket storage?

julianedwards replied:
Basically that this is not on the app or the application DB.

johndaniels commented on views.go:
```diff
@@ -201,43 +233,45 @@ func (lk *logKeeper) createTest(w http.ResponseWriter, r *http.Request) {
+			return
+		}
+	}
+	newTest.BuildName = build.Name
```
Can't we just move the creation of newTest down one line so that we can continue to use the object literal syntax to set BuildName?

julianedwards replied:
This code is to be deleted once we remove the DB writes, so I just reorganized it so it is simpler to remove it once we are ready. I am not opposed to moving it back though if you feel strongly.

johndaniels replied:
Ah, I see, because we plan on deleting buildname. That makes sense.

johndaniels commented on views.go:
```diff
@@ -201,43 +233,45 @@ func (lk *logKeeper) createTest(w http.ResponseWriter, r *http.Request) {
 		return
 	}
+	if build.S3 {
+		exists, err := lk.opts.Bucket.CheckMetadata(r.Context(), buildID, "")
```
This explicitly makes the behavior that if a build has "s3=true" but doesn't have the appropriate build metadata, we avoid doing the S3 upload/creation _and_ the DB one as well. Is that the intent here? One could also put this into the S3 parallel goroutine but it sounds like we want to just fail entirely at this point if the metadata isn't there?

julianedwards replied:
My assumption was that we would want to fail here if `s3=true` and the build does not exist in S3 since it would be invalid to have a test without a build. I did not put it in the goroutine because the code starts to get more messy than it already is since in some cases we want to return a 404 and others a 500.

johndaniels replied:
Ok. Wanted to confirm

johndaniels commented on storage/persistence.go:
```diff
@@ -59,3 +66,28 @@ func (b *Bucket) InsertLogChunks(ctx context.Context, buildID string, testID str
 }
+
+// CheckMetadata returns whether the metadata file exists for the given build
+// and (optional) test.
```
This comment implies it checks both, when it actually checks the test metadata specifically if the testID is supplied.

julianedwards replied:
Done.

johndaniels commented on storage/persistence.go:
```diff
@@ -59,3 +66,28 @@ func (b *Bucket) InsertLogChunks(ctx context.Context, buildID string, testID str
+
+// CheckMetadata returns whether the metadata file exists for the given build
+// and (optional) test.
+func (b *Bucket) CheckMetadata(ctx context.Context, buildID string, testID string) (bool, error) {
```
I know it's hard to test views.go, but we should probably at least have a small test for this function.

julianedwards replied:
Good catch, added tests!

julianedwards commented:
I moved `CheckMetadata` to `storage/retrieval.go` since it seemed like a more appropriate place. I also cleaned up some of the test code and added tests for `CheckMetadata`.

johndaniels commented on storage/storage_test.go:
```diff
@@ -10,17 +10,10 @@ import (
 	bucket, err := NewBucket(BucketOpts{
 		Location: PailLocal,
-		Path:     tempDir,
+		Path:     t.TempDir(),
```
Probably smart
